src/Ast.ml

```diff
@@ -852,6 +852,7 @@ end = struct
          |Pexp_construct
             ({txt= Lident "::"}, Some {pexp_desc= Pexp_tuple [_; e]})
          |Pexp_construct (_, Some e)
+         |Pexp_extension (_, PStr [{pstr_desc= Pstr_eval (e, _)}])
```

Comment:
I think that this is the source of the failed formatting you observe. This clause causes `continue` to be called on `e`, which leads to `sub_exp ~ctx:(Exp exp) e`. Calling `sub_exp` is effectively a (runtime) assertion that `e` is an immediate sub-expression of `exp`. The implementation of `sub_exp` checks this (in debug mode). The parenthesization code relies heavily on the property that 'subterms' are physically equal to an 'immediate' component value. Exactly what qualifies as 'immediate' is determined by the implementation of `check_exp` above. So in order to add this clause, an analogous one needs to be added to `check_exp`. I think that adding
```
| Pexp_extension (_, PStr [{pstr_desc= Pstr_eval (e, _)}]) -> assert (e == exp)
```
before line 413 should do it. Note that this is uncovering a collection of existing bugs, since `check_exp` currently assumes that there are no expression sub-terms of `Pexp_extension` nodes.

Reply:
I tried this and it didn't seem to fix the issue.  The change is in the latest commit though (
5bf3e38).

Reply:
I think this is fixed now thank you!